docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,132 @@
+---
+title: bmac-reboot-hosts-on-iso-change
+authors:
+  - "@mhrivnak"
+  - "@avishayt"
+  - "@flaper87"
+creation-date: 2021-06-30
+last-updated: 2021-07-01
+---
+
+# BMAC reboots only those hosts impacted by an ISO change
+
+## Summary
+
+When an InfraEnv's ISO gets regenerated based on a change such as adding or
+changing an NMStateConfig resource, there may be only a subset of that
+InfraEnv's Agents that are affected by the change. Rather than reboot all of
+them with the newly-generated ISO, only those hosts that are affected by the
+new or modified NMStateConfig should be rebooted.
+
+## Motivation
+
+If there are 500 NMStateConfigs and 500 Agents all related to the same
+InfraEnv, and then someone adds the 501st NMStateConfig, we do not want to
+reboot all 500 Agents. Rebooting is an expensive and time-consuming operation.
+We need a way to determine which Agents or hosts should be rebooted in response
+to an NMStateConfig.
+
+### Goals
+
+* Avoid rebooting un-provisioned BareMetalHosts when it is not needed.
+* Leave the door open for other platforms (include None/boot-it-yourself) to also avoid host reboots that are not needed.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+The assisted-service will track whether a particular Agent was booted using an
+old InfraEnv or NMStateConfig. It will ensure that any Agent that was booted
+using obsolete artifacts will have:
+* a label indicating that it was booted from obsolete artifacts
+* an array in its Status of references to the resources that are obsolete
+
+### assisted-service InfraEnv controller
+
+The InfraEnv ISO URL will change when the ISO contents change. This will make
+it obvious to an API user that they have booted a host from an obsolete ISO,
+because the URLs won't match. A side benefit of ensuring that the ISO URL
+uniquely identifies the resource's contents, is that pull-through caches will
+have an easy time understanding when an file is no longer relevant.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and compared against
```

Comment:
Maybe need to specify that it will happen only on registration right? 
What will be compared? does the service need to search for a matching nmstate config?

Reply:
Does it matter if it only happens on registration?

I'm adding a section below with the specific algorithm for adding and removing the label.

Reply:
adding this label will require going over all the nmstates and see if something changed not?
If user update nmstate and now it need to move to a different host or any other update
it may affect performance no?

Reply:
The agent will have in its status a reference to one specific NMStateConfig that it used. Then when reconciling the Agent, we only have to determine if that single resource changed or was removed.

Perhaps there's an edge case in what you're raising though. If a host is able to boot without a NMStateConfig and it creates an Agent resource, but then an NMStateConfig gets created for it, BMAC should recognize that and reboot the host. I'll add detail below to the BMAC behavior, and see what you think.

That should be a rare situation. Had the NMStateConfig resource been created at any time before the Agent resource appeared, then the host would have been restarted. 

Even if/when the controller does need to iterate all of the nmstateconfig resources, they're all in RAM anyway, so it shouldn't be a significant hit.

Reply:
i'm not concerned about executing a list, i'm thinking about that for every nmstate we need to unmarshal it and understand it before we create a link between agnet and nmstate.

Reply:
Assuming we are matching by MAC address, there is no unmarshalling required for [that field](https://github.com/openshift/assisted-service/blob/913aba6/internal/controller/api/v1beta1/nmstate_config_types.go#L52).